src/store.rs

```diff
@@ -1,12 +1,108 @@
 //! Store types.
 
+use std::sync::Arc;
+
+use matrix_sdk_crypto::store::{DynCryptoStore, IntoCryptoStore, MemoryStore};
 use wasm_bindgen::prelude::*;
 
 use crate::{
     encryption::EncryptionAlgorithm, identifiers::RoomId, impl_from_to_inner,
     vodozemac::Curve25519PublicKey,
 };
 
+/// A struct containing an open connection to a CryptoStore.
+///
+/// Opening the CryptoStore can take some time, due to the PBKDF calculation
+/// involved, so if multiple operations are being done on the same store, it is
+/// more efficient to open it once.
+#[wasm_bindgen]
+#[derive(Clone, Debug)]
+pub struct StoreHandle {
+    pub(crate) store: Arc<DynCryptoStore>,
+}
+
+#[wasm_bindgen]
+impl StoreHandle {
+    /// Open a crypto store.
+    ///
+    /// The created store will be based on IndexedDB if a `store_name` is
+    /// provided; otherwise it will be based on a memory store and once the
+    /// objects is dropped, the keys will be lost.
+    ///
+    /// # Arguments
+    ///
+    ///
+    /// * `store_name` - The name that should be used to open the IndexedDB
+    ///   based database. If this isn't provided, a memory-only store will be
+    ///   used. *Note* the memory-only store will lose your E2EE keys when the
+    ///   `StoreHandle` gets dropped.
+    ///
+    /// * `store_passphrase` - The passphrase that should be used to encrypt the
+    ///   store, for IndexedDB-based stores
+    #[wasm_bindgen(js_name = "open")]
+    pub async fn open_for_js(
+        store_name: Option<String>,
+        store_passphrase: Option<String>,
+    ) -> Result<JsValue, JsValue> {
+        Ok(StoreHandle::open(store_name, store_passphrase)
+            .await
+            .map_err(|e| JsError::from(&*e))?
+            .into())
+    }
```

Comment:
Ideally, you may want to split the Rust API from the JS API like so:

```rust
// The Rust API
impl StoreHandle {
    pub(crate) async fn open_(…) { … } // Note the name: `open_`
    // …
}

// The JS API
#[wasm_bindgen]
impl StoreHandle {
    pub async fn open(…) { … }
    // …
}
```

Reply:
@Hywan I don't quite follow. Are you asking that:
 * I put the two functions into separate `impl` blocks, with `#[wasm_bindgen]` on only one of them? If so, why is this an improvement?
 * I rename the functions? Happy to do so, but it feels very much like a matter of taste?

Reply:
Assuming this wasn't a blocking comment, so merging.